Sources/SecureXPC/Codable Property Wrappers/File Descriptor XPC Wrappers.swift

```diff
@@ -0,0 +1,184 @@
+    var descriptor: CInt { get }
+    var closeOnEncode: Bool { get }
+    func close() throws
+}
+
+private enum CodingKeys: String, CodingKey {
+    case descriptor
+    case closeOnEncode
+}
+
+extension FileDescriptorCodable {
+    public func encode(to encoder: Encoder) throws {
+        let container = try XPCEncoderImpl.asXPCEncoderImpl(encoder).xpcContainer(keyedBy: CodingKeys.self)
+        guard let encodedDescriptor = xpc_fd_create(self.descriptor) else {
+            let context = EncodingError.Context(codingPath: container.codingPath,
+                                                debugDescription: "Encoding failed for \(self.descriptor)",
+                                                underlyingError: nil)
+            throw EncodingError.invalidValue(self.descriptor, context)
+        }
+        
+        if self.closeOnEncode {
+            try self.close()
+        }
+        container.encode(encodedDescriptor, forKey: CodingKeys.descriptor)
+        try container.encode(self.closeOnEncode, forKey: CodingKeys.closeOnEncode)
+    }
+}
+
+extension FileDescriptorCodable {
+    public init(from decoder: Decoder) throws {
+        let container = try XPCDecoderImpl.asXPCDecoderImpl(decoder).xpcContainer(keyedBy: CodingKeys.self)
+        let descriptor = try container.decodeFileDescriptor(forKey: CodingKeys.descriptor)
+        // From xpc_fd_dup documentation: If the descriptor could not be created or if the given object was not an XPC
+        // file descriptor, -1 is returned.
+        if descriptor == -1 {
+            let context = DecodingError.Context(codingPath: container.codingPath,
+                                                debugDescription: "File descriptor could not be created",
+                                                underlyingError: nil)
+            throw DecodingError.dataCorrupted(context)
+        }
+        let closeOnEncode = try container.decode(Bool.self, forKey: CodingKeys.closeOnEncode)
+        
+        self = Self.init(descriptor: descriptor, closeOnEncode: closeOnEncode)
+    }
+}
+
+// MARK: Darwin file descriptor
+
+/// Wraps a Darwin file descriptor such that it can be sent over an XPC connection.
+///
+/// By default the provided file descriptor will be closed once it has been encoded.
+///
+/// When creating an ``XPCRoute`` that directly transfers this type as either the message or reply type, `DarwinFileDescriptorForXPC` must be the
+/// specified type, not `CInt`. This is not applicable when transferring a type which _contains_ a wrapped file descriptor as one of its properties.
+///
+/// This property wrapper shares an underlying representation with ``FileHandleForXPC`` and ``FileDescriptorForXPC`` and therefore may be used
+/// interchangeably between the server and client. However, to make use of such functionality requires routes with identical names and differing message and/or
+/// reply types.
+@propertyWrapper public struct DarwinFileDescriptorForXPC {
```

Comment:
What's a "Darwin file descriptor", as compared to a "normal" file descriptor of unix fame?

Reply:
NVM the darwin one *is* the unix one, whereas FileDescriptor is from the system framework. https://developer.apple.com/documentation/system/filedescriptor

Got it.

Reply:
Yeah that was the best name I could think of. If you have alternative recommendations I'd be happy to hear them. The other one I though of was "native" but preferred "Darwin" to that.

Reply:
POSIX sounds good!